Custom junit formatting

lorenzodalaqua opened the pull request:
This PRs customizes the title of a Scenario and other attributes when formatting test output using the `junit` format.

It mainly adds the line number and CircleCI node number to the scenario title.

EDIT: I ended up only adding any attributes if the `CIRCLE_NODE_INDEX` env variable is set. This way, tests in this project will not have to be affected, and I updated tests that use this patched version to set that variable when the added attributes are needed.

lorenzodalaqua committed:
Add line number to scenario name in Junit format

lorenzodalaqua committed:
Add file name to test case in Junit formatter

lorenzodalaqua committed:
Add CircleCI specific attributes to Junit formatter

lorenzodalaqua committed:
Remove example number from Scenario name in Junit formatter

This is applicable to Scenarios which are a part of a Scenario Outline.
Each example in an outline would be appended an index to its name in the
output (ex: #1, #2, #3...), but in a behat --rerun, they are all
appended #1, which breaks IFT detection. Not having the Example number
is the way the script currently expects the output to be, so it was
removed.

Chekote commented on src/Behat/Behat/Output/Node/Printer/JUnit/JUnitScenarioPrinter.php:
```diff
@@ -76,12 +76,63 @@ public function printOpenTag(Formatter $formatter, FeatureNode $feature, Scenari
+     * @return string
+     */
+    private function addLineNumberToScenarioName(string $name, int $lineNumber) {
+        return $name . " Line #" . $lineNumber;
```
Double quotes are necessary if you don't need string interpolation.
But I feel like if we use string interpolation, we can actually simplify this:

```suggestion
        return "$name Line #$lineNumber";
```

Please do this for similar string concatenation in this PR.

Chekote commented on src/Behat/Behat/Output/Node/Printer/JUnit/JUnitScenarioPrinter.php:
```diff
@@ -76,12 +76,63 @@ public function printOpenTag(Formatter $formatter, FeatureNode $feature, Scenari
+     */
+    private function addLineNumberToFeatureFile(string $path, int $lineNumber) {
+        return $path . ":" . $lineNumber;
+    }
```
Please put a blank line between methods.

Chekote commented on src/Behat/Behat/Output/Node/Printer/JUnit/JUnitScenarioPrinter.php:
```diff
@@ -76,12 +76,63 @@ public function printOpenTag(Formatter $formatter, FeatureNode $feature, Scenari
+     *
+     * This can be used to reference the scenario when running tests in the command line
+     *
+     * @param string $path
```
Let's either add a description to this or tweak the method description to be more obvious:

```suggestion
     * @param string $path the path to the feature file
```

```
     * Append the line number of the scenario within the feature file to the feature file's path
```

lorenzodalaqua replied:
I renamed this method and updated the comments to make it clearer what it is doing.

Chekote commented on src/Behat/Behat/Output/Node/Printer/JUnit/JUnitScenarioPrinter.php:
```diff
@@ -76,12 +76,63 @@ public function printOpenTag(Formatter $formatter, FeatureNode $feature, Scenari
+     */
+    private function addCircleCiAttributes(array $attributes) {
+        $circleCiNode = getenv('CIRCLE_NODE_INDEX');
+        if($circleCiNode !== false) {
```
```suggestion
        if ($circleCiNode !== false) {
```

Chekote commented on src/Behat/Behat/Output/Node/Printer/JUnit/JUnitScenarioPrinter.php:
```diff
@@ -76,12 +76,63 @@ public function printOpenTag(Formatter $formatter, FeatureNode $feature, Scenari
+    private function addCircleCiAttributes(array $attributes) {
+        $circleCiNode = getenv('CIRCLE_NODE_INDEX');
+        if($circleCiNode !== false) {
+            $circleCiNodeLabel =  "[Node #$circleCiNode] ";
+            $attributes['name'] = $circleCiNodeLabel . $attributes['name'];
```
The `$circleCiNodeLabel` variable seems unnecessary:

```suggestion
            $attributes['name'] = "[Node #$circleCiNode] ${attributes['name']}";
```

lorenzodalaqua committed:
Allow symfony/flex plugin globally

Chekote reviewed (changes requested):
I've also asked @BKorenek to weigh in on this and confirm that his concerns about navigating have been addressed.

lorenzodalaqua committed:
Remove line number from junit Scenario name output

lorenzodalaqua committed:
Remove circlecinodeindex attribute

lorenzodalaqua committed:
Remove uneeded variable.

lorenzodalaqua committed:
Use string interpolation instead of concatenation

lorenzodalaqua committed:
Improve name and docblock of method to append line number

BKorenek reviewed (approved):
I'm not super familiar with how this is all set up or exactly how it would translate, but at a glance it seems like ti would be good. The title is on its own from what I can tell, so it can be copied exactly and fed into `behat --name="<paste>"` and in addition the file and line number are dropped on there own line, pending donalds review being address it seems, which can also be used as `behat <paste>`. I'm a happy camper, as long as that copy button in CI picks up just the title, or the file with line number

lorenzodalaqua committed:
Only add file attribute in a CircleCI environment

Updated projects that use this modification to only rely on it if the
circleci environment variable is set.

Chekote commented on src/Behat/Behat/Output/Node/Printer/JUnit/JUnitScenarioPrinter.php:
```diff
@@ -98,7 +137,7 @@ private function buildExampleName(ExampleNode $scenario)
         }
 
-        $name = $currentOutline->getTitle() . ' #' . $this->outlineStepCount;
+        $name = $currentOutline->getTitle();
```
Why is this being removed? It seems to be part of Behat's normal behavior.

lorenzodalaqua replied:
Yeah, it is, I had forgotten about this modification. This is Behat normal behaviour, I added it back and will remove it only on CI.

This broke our automation script because when we run Behat a second time time with the `--rerun` flag all examples in a scenario outline would have `#1` in their title instead of the correct index. So if the second example failed in the first Behat run, in the rerun it would have a different title (e.g Scenario Title #2 would become Scenario Title #1 in the rerun).

I believe this happens because only the failing example is executed in the rerun and the way this method works is to keep a count of scenarios run with the same title, or maybe they are run in a different order in a rerun.

Chekote replied:
What do you think about this:

Instead of removing data from the normal Behat output, what amount moving this concern to the IFT automation script itself and leaving this as-is? The IFT automation script can strip the suffix off the end. That way, we can keep the modification of Behat to a minimum.

lorenzodalaqua replied:
Makes sense, I'll modify it in STDCheck and remove this here.

lorenzodalaqua committed:
Do not append example number in title on CI only

lorenzodalaqua committed:
Remove modification from buildExampleName()

Handle this use case in projects that use the patched version
